lib/kube/proxy/transport_retryable.go

```diff
@@ -0,0 +1,187 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/fs"
+	"log/slog"
+	"net/http"
+	"os"
+
+	"github.com/gravitational/trace"
+)
+
+// retryableTransport wraps an http.RoundTripper to enable automatic retry
+// on HTTP/2 GOAWAY frames.
+//
+// Kubernetes API servers send GOAWAY on HTTP/2 connections to redistribute
+// load across replicas. Go's http2.Transport automatically retries if
+// req.GetBody is set. Request bodies are buffered in memory for up to 5MB,
+// and buffered on disk beyond 5MB.
+type retryableTransport struct {
+	inner http.RoundTripper
+	log   *slog.Logger
+	ctx   context.Context
+}
+
+// RoundTrip implements http.RoundTripper and makes requests retryable.
+func (t *retryableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Skip streaming protocols which aren't retryable.
+	if t.isStreamingProtocol(req) {
+		return t.inner.RoundTrip(req)
+	}
+
+	// Retry is ok with no body, or buffered in call to GetBody.
+	if req.Body == nil || req.GetBody != nil {
+		return t.inner.RoundTrip(req)
+	}
+
+	// Make retryable by buffering and setting GetBody.
+	cleanup, err := t.makeRetryable(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if cleanup != nil {
+		defer func() {
+			// Cleanup may delete a temp file after roundtrip.
+			if err := cleanup(); err != nil && errors.Is(err, fs.ErrNotExist) {
+				if t.log != nil {
+					t.log.WarnContext(t.ctx, "Unable to cleanup temp file", "error", err)
+				}
+			}
+		}()
+	}
+
+	return t.inner.RoundTrip(req)
+}
+
+// makeRetryable buffers the request body to enable retry on HTTP/2 GOAWAY.
+func (t *retryableTransport) makeRetryable(req *http.Request) (func() error, error) {
+	// Skip if the request is already retryable
+	if req.Body == nil || req.GetBody != nil {
+		return nil, nil
+	}
+
+	// Limit the body buffer size to a large 5X for most cases.
+	// ConfigMaps and Secrets have 1MB limits with kube etcd.
+	// Large bodies might come from a custom CRD or batch operation,
+	// and is likely rare.
+	const maxBufferSize = 5 * 1024 * 1024 // 5MB
```

Comment:
We are not limiting the body. this is a threshold for us to choose between the memory vs disk. please update the comment